### PR TITLE
[LLVM][DWARF] Fix accelerator table switching between CU and TU

### PR DESCRIPTION
**Pull request by ayermolo**
Bug 1 is triggered when a TU is already created, and we process the same
DICompositeType at a top level. We would switch to TU accelerator table, but
would not switch back on early exit. As the result we would add CU entries to the TU
accelerator table. When we try to write out TUs and normalize entries, the
offsets for DIEs that are part of a CU would not have been computed, and it
would assert on getOffset().

Bug 2 is triggered when processing nested TUs. When we exit from addDwarfTypeUnitType we switched back to CU accelerator table. If we were processing nested TUs, the rest of the entries from TUs would be added to CU accelerator table. When we write out TUs, all the DIE pointers will become invalid. Eventually it will assert during normalization step after CU is processed.

### CODE REVIEW

**Comment by ayermolo** [comment]
There is another bug on when we exit from addDwarfTypeUnitType. I need to construct a small repro for it.

File: llvm/lib/CodeGen/AsmPrinter/DwarfDebug.h [hunk 1]
```diff
@@ -925,6 +926,7 @@ class DwarfDebug : public DebugHandlerBase {
 
   /// Sets the current DWARF5AccelTable to use.
   void setCurrentDWARF5AccelTable(const DWARF5AccelTableKind Kind) {
+    CurrentKind = Kind;
```
Comment (dwblaikie) [inline llvm/lib/CodeGen/AsmPrinter/DwarfDebug.h:929]: Could skip this variable and check `CurrentDebugNames == &AccelDebugNames` or similar.

**Review by dwblaikie — APPROVED** [review]
Looks OK but please skip the extra assert-only `CurrentKind` variable in favor of testing the `CurrentDebugNames` value.

**Comment by vitalybuka** [comment]
Could this be caused by the patch https://lab.llvm.org/buildbot/#/builders/236/builds/8673 ?

**Comment by ayermolo** [comment]
> Could this be caused by the patch https://lab.llvm.org/buildbot/#/builders/236/builds/8673 ?

Passes for me locally. Also patch impacts .debug_names for DWARF. So highly unlikely it would cause this test to fail.

**Comment by rorth** [comment]
However, the patch broke the [Solaris/sparcv9 buildbot](https://lab.llvm.org/buildbot/#/builders/72/builds/1834):
```
llc: error: unable to get target for 'x86_64-unknown-linux-gnu', see --version and --triple.
```
The bot is configured to do a Sparc-only build.

**Comment by ayermolo** [comment]
> However, the patch broke the [Solaris/sparcv9 buildbot](https://lab.llvm.org/buildbot/#/builders/72/builds/1834):
> 
> ```
> llc: error: unable to get target for 'x86_64-unknown-linux-gnu', see --version and --triple.
> ```
> 
> The bot is configured to do a Sparc-only build.

oh. I think 
; REQUIRES: x86-registered-target is missing. Let me add it

**Pull request merged**